File: ALittle/Module/ALittleIDE/Other/GameLibrary/Std/Script/Std/Loop/LoopFunction.lua

module("ALittle", package.seeall)

local ___rawset = rawset
local ___pairs = pairs
local ___ipairs = ipairs


assert(ALittle.LoopObject, " extends class:ALittle.LoopObject is nil")
LoopFunction = Lua.Class(ALittle.LoopObject, "ALittle.LoopFunction")

function LoopFunction:Ctor(func, count, interval_time, delay_time)
	___rawset(self, "_func", func)
	___rawset(self, "_total_count", count)
	___rawset(self, "_force_completed", false)
	if delay_time == nil or delay_time < 0 then
		delay_time = 0
	end
	if interval_time < 0 then
		interval_time = 0
	end
	___rawset(self, "_total_interval_time", interval_time)
	___rawset(self, "_total_delay_time", delay_time)
	___rawset(self, "_accumulate_count", 0)
	___rawset(self, "_accumulate_time", 0)
	___rawset(self, "_accumulate_delay_time", 0)
	if self._func == nil then
		___rawset(self, "_accumulate_count", 0)
		___rawset(self, "_total_count", 0)
		Log("LoopFunction create failed:function is nil or not a function")
		return
	end
	___rawset(self, "_complete_callback", nil)
end

function LoopFunction.__getter:complete_callback()
	return self._complete_callback
end

function LoopFunction.__setter:complete_callback(value)
	self._complete_callback = value
end

function LoopFunction:Reset()
	self._accumulate_count = 0
	self._accumulate_time = 0
	self._accumulate_delay_time = 0
end

function LoopFunction:Completed()
	if self._complete_callback ~= nil then
		self._complete_callback()
	end
end

function LoopFunction:IsCompleted()
	if self._force_completed then
		return true
	end
	if self._total_count < 0 then
		return false
	end
	return self._total_count <= self._accumulate_count
end

function LoopFunction:SetCompleted()
	if self._total_count < 0 then
		self._force_completed = true
		return
	end
	self._accumulate_count = self._total_count
end

function LoopFunction:SetTime(time)
	self._accumulate_count = 0
	self._accumulate_delay_time = 0
	self._accumulate_time = 0
	if time <= 0 then
		return 0, false
	end
	if self._total_count < 0 then
		return 0, false
	end
	if time <= self._total_delay_time then
		self._accumulate_delay_time = time
		return 0, false
	end
	self._accumulate_delay_time = self._total_delay_time
	time = time - (self._total_delay_time)
	local total_time = self._total_interval_time * self._total_count
	if time < total_time then
		local float_count = total_time / self._total_interval_time
		local count = Math_Floor(float_count)
		self._accumulate_count = count
		self._accumulate_time = Math_Floor((float_count - count) * self._total_interval_time)
		return 0, false
	end
	self._accumulate_count = self._total_count
	return time - total_time, true
end

function LoopFunction:Update(frame_time)
	if self._accumulate_delay_time < self._total_delay_time then
		self._accumulate_delay_time = self._accumulate_delay_time + (frame_time)
		if self._accumulate_delay_time < self._total_delay_time then
			return
		end
		frame_time = self._accumulate_delay_time - self._total_delay_time
		self._accumulate_delay_time = self._total_delay_time
	end
	self._accumulate_time = self._accumulate_time + (frame_time)
	if self._accumulate_time > self._total_interval_time then
		self._accumulate_time = 0
		self._accumulate_count = self._accumulate_count + 1
		self._func()
	end
end

File: ALittle/Module/ALittleIDE/Other/GameLibrary/Std/JSScript/Std/Loop/LoopFunction.js
{
if (typeof ALittle === "undefined") window.ALittle = {};


if (ALittle.LoopObject === undefined) throw new Error(" extends class:ALittle.LoopObject is undefined");
ALittle.LoopFunction = JavaScript.Class(ALittle.LoopObject, {
	Ctor : function(func, count, interval_time, delay_time) {
		this._func = func;
		this._total_count = count;
		this._force_completed = false;
		if (delay_time === undefined || delay_time < 0) {
			delay_time = 0;
		}
		if (interval_time < 0) {
			interval_time = 0;
		}
		this._total_interval_time = interval_time;
		this._total_delay_time = delay_time;
		this._accumulate_count = 0;
		this._accumulate_time = 0;
		this._accumulate_delay_time = 0;
		if (this._func === undefined) {
			this._accumulate_count = 0;
			this._total_count = 0;
			ALittle.Log("LoopFunction create failed:function is nil or not a function");
			return;
		}
		this._complete_callback = undefined;
	},
	get complete_callback() {
		return this._complete_callback;
	},
	set complete_callback(value) {
		this._complete_callback = value;
	},
	Reset : function() {
		this._accumulate_count = 0;
		this._accumulate_time = 0;
		this._accumulate_delay_time = 0;
	},
	Completed : function() {
		if (this._complete_callback !== undefined) {
			this._complete_callback();
		}
	},
	IsCompleted : function() {
		if (this._force_completed) {
			return true;
		}
		if (this._total_count < 0) {
			return false;
		}
		return this._total_count <= this._accumulate_count;
	},
	SetCompleted : function() {
		if (this._total_count < 0) {
			this._force_completed = true;
			return;
		}
		this._accumulate_count = this._total_count;
	},
	SetTime : function(time) {
		this._accumulate_count = 0;
		this._accumulate_delay_time = 0;
		this._accumulate_time = 0;
		if (time <= 0) {
			return [0, false];
		}
		if (this._total_count < 0) {
			return [0, false];
		}
		if (time <= this._total_delay_time) {
			this._accumulate_delay_time = time;
			return [0, false];
		}
		this._accumulate_delay_time = this._total_delay_time;
		time = time - (this._total_delay_time);
		let total_time = this._total_interval_time * this._total_count;
		if (time < total_time) {
			let float_count = total_time / this._total_interval_time;
			let count = ALittle.Math_Floor(float_count);
			this._accumulate_count = count;
			this._accumulate_time = ALittle.Math_Floor((float_count - count) * this._total_interval_time);
			return [0, false];
		}
		this._accumulate_count = this._total_count;
		return [time - total_time, true];
	},
	Update : function(frame_time) {
		if (this._accumulate_delay_time < this._total_delay_time) {
			this._accumulate_delay_time = this._accumulate_delay_time + (frame_time);
			if (this._accumulate_delay_time < this._total_delay_time) {
				return;
			}
			frame_time = this._accumulate_delay_time - this._total_delay_time;
			this._accumulate_delay_time = this._total_delay_time;
		}
		this._accumulate_time = this._accumulate_time + (frame_time);
		if (this._accumulate_time > this._total_interval_time) {
			this._accumulate_time = 0;
			++ this._accumulate_count;
			this._func();
		}
	},
}, "ALittle.LoopFunction");

}